[Merge] add merge Beacon API support

tbenr opened the pull request:
## PR Description

add Merge support in REST beacon APIs

related to #4504

- reorganize `SignedBeaconBlock` and `BeaconBlock*`

## Documentation

- [ ] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

tbenr committed:
add merge api support

tbenr committed:
fixes AT

tbenr committed:
reorganize SignedBeaconBlock

tbenr committed:
c&p issue

tbenr committed:
improve UT

tbenr committed:
fix UT compilation

tbenr committed:
improvements

tbenr committed:
BeaconBlock to BeaconBlockPhase0
improvement in BeaconBlock* constructors

rolfyone commented:
it's looking pretty good!

the merge types need a whole lot of schema annotations to illustrate their types. without it, they come out like:
![Screen Shot 2021-11-01 at 10 22 20 am](https://user-images.githubusercontent.com/2967240/139606370-d112eb1f-8de4-4425-816a-08c99a447738.png)

In most other areas you'll see that they're basically 'string' type, defined by schema manually on each field...

also in beacon state, the enum updated but there's no merge `state`
![Screen Shot 2021-11-01 at 10 17 56 am](https://user-images.githubusercontent.com/2967240/139606426-a0dbc5c3-aabf-4012-b3ed-444664fc0fe7.png)

tbenr committed:
add ExecutionPayload api schema annotations
add BeaconStateMerge in State schema annotation
add BeaconBlockMerge in UnsignedBlock schema annotation
fix BeaconStateAltair schema properties names

tbenr committed:
add ExecutionPayload api schema annotations
add BeaconStateMerge in State schema annotation
add BeaconBlockMerge in UnsignedBlock schema annotation
fix BeaconStateAltair schema properties names
add UInt256 serializer\deserializer
expand BeaconState api UT

tbenr committed:
fix UT

rolfyone commented on data/serializer/src/main/java/tech/pegasys/teku/api/schema/BeaconBlock.java:
```diff
@@ -30,39 +28,38 @@
 
   @Schema(type = "string", format = "uint64")
-  public final UInt64 proposer_index;
+  public final UInt64 proposerIndex;
```
so by doing this, we effectively create some duplication - now the interface isn't sure about whether its `proposer_index` or `proposerIndex`, so we have both.
Need to at least put `@JsonProperty("proposer_index")` to define it. this will also get rid of the typing issue appearing because of the proposer_index type not being given a schema definition.
same goes for all renamed fields here.

rolfyone replied:
the other option is leave them named the way the api requires them, given that it's only an object that the api uses...

rolfyone commented on data/serializer/src/main/java/tech/pegasys/teku/api/schema/merge/BeaconStateMerge.java:
```diff
@@ -0,0 +1,168 @@
+  @JsonProperty("latest_execution_payload_header")
+  public final ExecutionPayloadHeader latestExecutionPayloadHeader;
+
+  public BeaconStateMerge(
```
should have @JSonCreator at the start as the properties below are only set because of it being a creator...

the genesis_validators_root property doesnt look like its come through correctly, its coming up with 'empty' and 'zero' attributes in the swagger on the altair state, so that suggests it needs an annotation (`@Schema`)

Also the merge state isn't listed, so its missing from the `oneOf` in BeaconState.

tbenr replied:
i think `genesis_validators_root` looks not good in master too, fixing :)
going to check the @JsonCreator
the missing `oneOf` is actually not missing... you might hadn't rebuild the latest version?

rolfyone commented on data/serializer/src/main/java/tech/pegasys/teku/api/schema/altair/BeaconStateAltair.java:
```diff
@@ -40,82 +40,89 @@
+        previousJustifiedCheckpoint,
+        currentJustifiedCheckpoint,
+        finalizedCheckpoint);
+    this.previousEpochParticipation = previousEpochParticipation;
+    this.currentEpochParticipation = currentEpochParticipation;
+    this.inactivityScores = inactivityScores;
+    this.currentSyncCommittee = currentSyncCommittee;
+    this.nextSyncCommittee = nextSyncCommittee;
```
kind of a repeat, but basically all these fields are now incorrect in the api returns, which is why I'm suggesting putting their names back unless we have a good reason.

rolfyone commented on data/serializer/src/main/java/tech/pegasys/teku/api/schema/merge/ExecutionPayloadCommon.java:
```diff
@@ -0,0 +1,206 @@
+
+public abstract class ExecutionPayloadCommon {
+
+  @Schema(
+      name = "parent_hash",
+      type = "string",
+      format = "byte",
+      pattern = PATTERN_BYTES32,
+      description = DESCRIPTION_BYTES32)
+  public final Bytes32 parentHash;
+
```
i think all of these have come out incorrectly because of the abstract class usage. i'm not sure how you'd get around that, but the swagger is all busted for this class...

tbenr replied:
I guess you are not seeing the latest version, can you do a clean rebuild?

tbenr committed:
add Bytes20 serializer-deserializer

tbenr committed:
fix BeaconBlock schema names

tbenr committed:
update with the latest master

tbenr committed:
fix genesis_validators_root and missing @JsonCreator on BeaconStateMerge, ExecutionPayload and ExecutionPayloadHeader

tbenr commented:
getState API
<img width="1844" alt="Schermata 2021-11-04 alle 16 21 44" src="https://user-images.githubusercontent.com/15999009/140350401-3089afa6-c427-40e6-a83c-e245ea501674.png">

getBlock API
<img width="2239" alt="Schermata 2021-11-04 alle 16 26 17" src="https://user-images.githubusercontent.com/15999009/140353415-9b16d559-0b3c-45c5-97e3-327c6eb9325f.png">

@rolfyone to me looks ok now..hopefully :-)

tbenr committed:
remove type field in SignedBeaconBlock

tbenr committed:
remove type field in SignedBeaconBlock

tbenr committed:
fix deserialization

tbenr committed:
fix rebase

rolfyone reviewed (approved):
I wasn't able to easily run it today, just had a messed up environment of my own, but from the screenshots it looks like exactly what we talked about - looks good!

ajsutton committed:
Ensure we get Transaction subtypes appropriately.

tbenr committed:
Merge branch 'master' into merge_beacon_apis

tbenr committed:
fix rebase
